java/arcs/core/data/expression/InferredType.kt

```diff
@@ -0,0 +1,133 @@
+  open fun isAssignableFrom(other: InferredType): Boolean = this == other
+  fun union(other: InferredType) = UnionType(setOf(this, other))
+
+  /** Represents all primitive types such as numbers, text, booleans, etc. */
+  sealed class Primitive(val kClass: KClass<*>) : InferredType() {
+    override fun isAssignableFrom(other: InferredType): Boolean = when {
+      this is Numeric && other is Numeric -> widen(this, other) == this
+      else -> this == other
+    }
+
+    override fun toString() = "PrimitiveType(${kClass.simpleName ?: super.toString()})"
+
+    object DoubleType : Primitive(Double::class), Numeric
+    object FloatType : Primitive(Float::class), Numeric
+    object BigIntType : Primitive(BigInt::class), Numeric
+    object LongType : Primitive(Long::class), Numeric
+    object IntType : Primitive(Int::class), Numeric
+    object ShortType : Primitive(Short::class), Numeric
+    object ByteType : Primitive(Byte::class), Numeric
+    object NumberType : Primitive(Number::class), Numeric
+    object BooleanType : Primitive(Boolean::class)
+    object TextType : Primitive(String::class)
+    object NullType : Primitive(NullType::class)
+
+    companion object {
+      private val kclassToType by lazy {
+        listOf(
+          DoubleType,
+          FloatType,
+          BigIntType,
+          LongType,
+          IntType,
+          ShortType,
+          ByteType,
+          NumberType,
+          BooleanType,
+          TextType,
+          NullType
+        ).associateBy({ it.kClass }, { it })
+      }
+
+      /** All primitive types. */
+      val allTypes by lazy {
+        kclassToType.values.toSet()
+      }
+
+      /** Given a [KClass] return the corresponding [InferredType]. */
+      fun of(kClass: KClass<*>) = requireNotNull(kclassToType[kClass]) {
+        "Unable to map $kClass to an InferredType."
+      }
+    }
+  }
+
+  /** Represents union of several possible types. */
+  data class UnionType(val types: Set<InferredType>) : InferredType() {
+    constructor(vararg types: InferredType) : this(setOf(*types))
+
+    override fun toString() =
+      "UnionType${types.joinToString(prefix = "(", postfix = ")", separator = ", ")}"
+
+    override fun isAssignableFrom(other: InferredType): Boolean = when (other) {
+      is UnionType -> other.types.all { this.isAssignableFrom(it) }
+      else -> types.any { it.isAssignableFrom(other) }
+    }
+  }
+
+  /** Represents a seequence of values of the given type. */
+  data class SeqType(val type: InferredType) : InferredType() {
+    override fun toString() = "SeqType($type)"
+    override fun isAssignableFrom(other: InferredType): Boolean = when (other) {
+      is SeqType -> type.isAssignableFrom(other.type)
+      else -> false
+    }
+  }
+
+  /** Represents a type corresponding to a scope with variables of various types. */
+  data class ScopeType(val scope: Expression.Scope) : InferredType() {
+    override fun toString() = "ScopeType(${scope.scopeName})"
```

Comment:
Not worth printing field names and types?

If I wrote for example: `new Foo {abc: "Yay", def: 42}.abd` and the error told me:
`Field "abd" not found on type "Foo {abc: Text, def: Int}"`
that would be more helpful than:
`Field "abd" not found on type "Foo"`.
(Though I would still manage to figure it out with the second error message).

WDYT?